File: can/generate_can_headers.py
import cantools
import argparse
import re
from pathlib import Path


def sanitize_name(name: str) -> str:
    """Sanitize a string to be a valid C identifier."""
    name = re.sub(r'\W+', '_', name)
    if name and name[0].isdigit():
        name = "_" + name
    return name


def c_type_for_signal(signal):
    length = signal.length

    if signal.is_float:
        return "float"

    if signal.is_signed:
        if length <= 8:
            return "int8_t"
        elif length <= 16:
            return "int16_t"
        elif length <= 32:
            return "int32_t"
        else:
            return "int64_t"
    else:
        if length <= 8:
            return "uint8_t"
        elif length <= 16:
            return "uint16_t"
        elif length <= 32:
            return "uint32_t"
        else:
            return "uint64_t"


def generate_enum(signal, message_name):
    enum_name = sanitize_name(f"{message_name}_{signal.name}_e".lower())

    lines = []
    lines.append("typedef enum {")

    for value, choice in signal.choices.items():
        if hasattr(choice, "name"):
            text = choice.name
        else:
            text = str(choice)

        member = sanitize_name(
            f"{message_name}_{signal.name}_{text}".upper()
        )
        lines.append(f"    {member} = {value},")

    lines.append(f"}} {enum_name};\n")
    return "\n".join(lines)


def generate_struct(message):
    struct_name = sanitize_name(message.name.lower() + "_t")

    lines = []
    lines.append("typedef struct {")

    for signal in message.signals:
        ctype = c_type_for_signal(signal)
        member = sanitize_name(signal.name)
        lines.append(f"    {ctype} {member};")

    lines.append(f"}} {struct_name};\n")
    return "\n".join(lines)


def generate_header(dbc_path: Path, output_file: Path | None):
    db = cantools.database.load_file(str(dbc_path))

    if output_file is None:
        output_file = dbc_path.with_name(f"{dbc_path.stem}_can_msgs.h")

    can_id_macros = []
    can_len_macros = []
    enums = []
    structs = []

    # Sort messages by numeric CAN ID
    sorted_messages = sorted(db.messages, key=lambda m: m.frame_id)

    for message in sorted_messages:
        msg_macro_name = sanitize_name(message.name.upper())

        # CAN ID macro
        can_id_macros.append(
            f"#define CAN_ID_{msg_macro_name} 0x{message.frame_id:X}"
        )

        # CAN length macro
        can_len_macros.append(
            f"#define CAN_DLC_{msg_macro_name} {message.length}"
        )

        for signal in message.signals:
            if signal.choices:
                enums.append(generate_enum(signal, message.name.lower()))

        structs.append(generate_struct(message))

    with open(output_file, "w") as f:
        f.write("#pragma once\n\n")
        f.write("#include <stdint.h>\n\n")

        f.write("/* ================= CAN ID Macros ================= */\n\n")
        for macro in can_id_macros:
            f.write(macro + "\n")

        f.write("\n/* ================= CAN Length Macros ================= */\n\n")
        for macro in can_len_macros:
            f.write(macro + "\n")

        f.write("\n\n")

        if enums:
            f.write("/* ================= Value Table Enums ================= */\n\n")
            for enum_code in enums:
                f.write(enum_code + "\n")

        f.write("/* ================= Message Structs ================= */\n\n")
        for struct_code in structs:
            f.write(struct_code + "\n")

    print(f"Generated: {output_file}")


def main():
    parser = argparse.ArgumentParser(
        description="Generate C header files from a CAN DBC file."
    )

    parser.add_argument(
        "dbc",
        type=Path,
        help="Input DBC file"
    )

    parser.add_argument(
        "-o",
        "--output",
        type=Path,
        help="Output header filename (optional)"
    )

    args = parser.parse_args()

    if not args.dbc.is_file():
        parser.error(f"DBC file not found: {args.dbc}")

    generate_header(args.dbc, args.output)


if __name__ == "__main__":
    main()

File: can/mdc/tools/generate_can_headers.mjs
#!/usr/bin/env node
/**
 * generate_can_headers — generate a C header from an MDC v3 project.
 *
 * Emits CAN id + DLC macros, value-table enums, and per-message structs.
 *
 * Usage: node tools/generate_can_headers.mjs <project-dir|project.mdc.json> [-o out.h]
 */
import { writeFile } from "node:fs/promises";
import { loadProject, iterMessages, resolveChoices } from "../lib/mdc-load.mjs";

const args = process.argv.slice(2);
const projectPath = args[0];
const outIdx = args.findIndex((a) => a === "-o" || a === "--output");
const outFile = outIdx >= 0 ? args[outIdx + 1] : null;
if (!projectPath) {
  console.error("Usage: node tools/generate_can_headers.mjs <project-dir|project.mdc.json> [-o out.h]");
  process.exit(2);
}

const sanitize = (name) => {
  let s = String(name).replace(/\W+/g, "_");
  if (/^[0-9]/.test(s)) s = "_" + s;
  return s;
};

function cType(sig) {
  if (sig.is_float) return sig.length > 32 ? "double" : "float";
  const signed = sig.is_signed;
  const w = sig.length;
  const base = w <= 8 ? "8" : w <= 16 ? "16" : w <= 32 ? "32" : "64";
  return `${signed ? "int" : "uint"}${base}_t`;
}

const project = await loadProject(projectPath);

const idMacros = [];
const lenMacros = [];
const enums = [];
const structs = [];

const messages = [...iterMessages(project)].sort((a, b) => a.message.frame_id - b.message.frame_id);
for (const { network, message } of messages) {
  const macro = sanitize(message.name.toUpperCase());
  idMacros.push(`#define CAN_ID_${macro} 0x${message.frame_id.toString(16).toUpperCase()}`);
  lenMacros.push(`#define CAN_DLC_${macro} ${message.length}`);

  const structName = sanitize(message.name.toLowerCase() + "_t");
  const members = [];
  for (const sig of message.signals ?? []) {
    members.push(`    ${cType(sig)} ${sanitize(sig.name)};`);
    const choices = resolveChoices(sig, network, project);
    if (choices?.length) {
      const enumName = sanitize(`${message.name}_${sig.name}_e`.toLowerCase());
      const lines = ["typedef enum {"];
      for (const c of choices) {
        const member = sanitize(`${message.name}_${sig.name}_${c.label}`.toUpperCase());
        lines.push(`    ${member} = ${c.value},`);
      }
      lines.push(`} ${enumName};`);
      enums.push(lines.join("\n"));
    }
  }
  structs.push(["typedef struct {", ...members, `} ${structName};`].join("\n"));
}

const out = [];
out.push("#pragma once");
out.push("");
out.push("#include <stdint.h>");
out.push("");
out.push("/* ================= CAN ID Macros ================= */");
out.push("");
out.push(...idMacros);
out.push("");
out.push("/* ================= CAN Length Macros ================= */");
out.push("");
out.push(...lenMacros);
out.push("");
if (enums.length) {
  out.push("/* ================= Value Table Enums ================= */");
  out.push("");
  out.push(enums.join("\n\n"));
  out.push("");
}
out.push("/* ================= Message Structs ================= */");
out.push("");
out.push(structs.join("\n\n"));
out.push("");

const text = out.join("\n");
if (outFile) {
  await writeFile(outFile, text);
  console.error(`Generated: ${outFile}`);
} else {
  process.stdout.write(text);
}
